### PR TITLE
[JIT] Add special cases batch_norm, instance_norm and dropout for SchemaInfo

### PR DESCRIPTION
**Pull request by ttanpcs**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81518
* #81444
* #81439
* #81352
* __->__ #81007

- Added special cases for detach in is_non_deterministic() check and batch_norm and instance_norm in is_mutable() check in SchemaInfo().
- Added tests for the above special cases for detach, batch_norm and instance_norm.

### CODE REVIEW

File: torch/csrc/utils/schema_info.cpp [hunk 1]
```diff
@@ -67,7 +67,26 @@ bool SchemaInfo::is_mutable(size_t index) {
   return std::any_of(
       input_alias_map_[index].begin(),
       input_alias_map_[index].end(),
-      [this](size_t index) { return this->schema_.is_mutable(index); });
+      [this](size_t index) {
+        static const char* batch_norm_literal =
```
Comment (davidberard98) [inline torch/csrc/utils/schema_info.cpp:71]: A few suggestions here:

1. Can you do the schema parsing statically so that it doesn't have to happen each time?
2. Can you move the special case behavior outside of the `std::any_of`?  The special case behavior doesn't really have any dependence on index, it would probably be easier to understand just as `if (this_schema == batch_norm_literal && (has_running_mean || has_running_var) { return training_is_true; }`

**Review by davidberard98 — COMMENTED** [review]
Nice, it's good to see the special cases in here! Just had some comments about how to handle the special cases in SchemaInfo

File: torch/csrc/utils/schema_info.cpp (hunk 1 above)
Comment (goldenxuett) [reply torch/csrc/utils/schema_info.cpp:71]: 1. Yeah I'll clean that up and also make the special case look more readable.
2. Technically there actually is a dependence on index. For instance, with batch_norm if the input tensor value is aliasing the running_mean and training = true, then the input tensor will now be mutable which is found through std::any_of

File: torch/csrc/utils/schema_info.cpp [hunk 2]
```diff
@@ -67,10 +67,34 @@ bool SchemaInfo::is_mutable(size_t index) {
...
+      "aten::instance_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool use_input_stats, float momentum, float eps, bool cudnn_enabled) -> Tensor");
+
+  // Note that the batch norm and instance norm checks depend on index because
+  // of cases where either running_mean or running_var alias another input
+  // argument causing its alias status to change.
   return std::any_of(
       input_alias_map_[index].begin(),
       input_alias_map_[index].end(),
-      [this](size_t index) { return this->schema_.is_mutable(index); });
+      [this](size_t index) {
```
Comment (davidberard98) [inline torch/csrc/utils/schema_info.cpp:82]: sorry, one more request I didn't notice last time - could you rename this from `index` to something that doesn't shadow a previous variable? (actually I got confused in the last review and that's why I had the previous suggestion).

**Review by davidberard98 — CHANGES_REQUESTED** [review]

**Review by davidberard98 — APPROVED** [review]
nice, looks good!
